BlueprintUI/Tests/HintingSizeCacheTests.swift

```diff
@@ -0,0 +1,166 @@
+import XCTest
+@testable import BlueprintUI
+
+final class HintingSizeCacheTests: XCTestCase {
+
+    func test_caching() {
+
+        func test(key: SizeConstraint) {
+            let size = CGSize(width: 100, height: 100)
+
+            let boundaries = key.boundaries(betweenSelfAndSize: size)
+            // test all the keys twice
+            let keys = boundaries + boundaries
+
+            assertMisses(
+                keys: keys,
+                size: size,
+                // we expect to hit each of the unique keys exactly once
+                misses: boundaries,
+                options: .init(hintRangeBoundaries: false, searchUnconstrainedKeys: false)
+            )
+        }
+
+        test(key: SizeConstraint(width: .atMost(200), height: .atMost(200)))
+        test(key: SizeConstraint(width: .atMost(200), height: .unconstrained))
+        test(key: SizeConstraint(width: .unconstrained, height: .atMost(200)))
+        test(key: .unconstrained)
+    }
+
+    func test_hintRangeBoundaries() throws {
+
+        func test(key: SizeConstraint) {
+            let size = CGSize(width: 100, height: 100)
+
+            let boundaries = key.boundaries(betweenSelfAndSize: size)
+            let offKeys = [
+                SizeConstraint(width: key.width, height: .atMost(150)),
+                SizeConstraint(width: .atMost(150), height: key.height),
+            ]
+            // test all the keys twice
+            let keys = boundaries + boundaries + offKeys + offKeys
```

Comment:
Can I read this as "Visit all keys twice. The first visit will be a miss, the second should be a hit if it was hinted by the first."